### ISSUE TITLE
inZone call shoud not create task

### ISSUE THREAD

**Issue by amaurel** [tekartik/synchronized.dart#1]
Here is a test case that fails
```dart
       var value = '';

        Future outer1 = synchronized('test', () async {

          expect(value, equals(''));
          value = 'outer1';

          await sleep(20);

          await synchronized('test', () async {
            await sleep(30);
            expect(value, equals('outer1'));
            value = 'inner1';
          });
        });

        Future outer2 = synchronized('test', () async {
          await sleep(30);
          expect(value, equals('inner1'));
          value = 'outer2';
        });

        Future outer3 = sleep(30).whenComplete((){
          return synchronized('test', () async {
            expect(value, equals('outer2'));
            value = 'outer3';
          });
        });

        await Future.wait([outer1, outer2, outer3]);

        expect(value, equals('outer3'));
```

**Comment by amaurel** [comment]
not creating task for inZone calls resolve the problem

```dart
  // implementation
  Future/*<T>*/ synchronized/*<T>*/(computation(), {timeout: null}) {
    // Inner case scenario
//    //we don't add task if we are in zone
    if (inZone) {
      return new Future.sync(() {
        if (computation != null) {
          return computation();
        }
      });
    }

...
...

```

**Comment by alextekartik** [comment]
Good catch, indeed i'm not dealing with the inner task properly. Looking at it right now as I found other issues even with the fix you propose. thanks

**Comment by alextekartik** [comment]
I add you sample as a new unit test, thanks. The issue is now fixed. There were other issues related to inner tasks that needed a more complete fixes than what you propose (i.e. it will always wait for the inner tasks to complete before scheduling a new one). Let me know if it fixes the issue you might have encountered in other scenarios

**Comment by amaurel** [comment]
I am curious, in which scenario do you need a task to wait for its inner tasks to complete ?

**Comment by alextekartik** [comment]
Actually the synchronized block itself won't wait for the inner synchronized blocks to finish. There was a bug (that your sample shows) where a new outer synchronized block was not waiting for previous inner synchronized block (in another outer block) to be complete and actually the whole implementation was wrong on that. Is is fixed in 1.1.0. In the code a SynchronizedTask groups an outer block and its inner blocks so that the next SynchronizedTask is ran after all of them are complete. Note sure I'm clear very on that. I'm pretty sure a better implementation could be done especially when the timing for async functions start changes (https://github.com/dart-lang/sdk/blob/master/docs/newsletter/20170915.md). Eagerly waiting for dart 2.0 for that
